Source/CombatExtended/CombatExtended/Comps/CompAmmosetSwitcher.cs

```diff
@@ -0,0 +1,345 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using RimWorld;
+using Verse;
+using UnityEngine;
+using CombatExtended.Utilities;
+
+namespace CombatExtended
+{
+    public class CompUnderBarrel : CompRangedGizmoGiver
+    {
+        public CompProperties_UnderBarrel Props => (CompProperties_UnderBarrel)this.props;
+
+        public CompEquippable CompEq() => this.parent.TryGetComp<CompEquippable>();
+        
+        public CompEquippable _compEq;
+        
+        public CompEquippable compEq
+        {
+            get
+            {
+                if(_compEq == null)
+                {
+                    _compEq = CompEq();
+                }
+                
+                return _compEq;
+            }
+        }
+
+        public CompAmmoUser CompAmmo() => this.parent.TryGetComp<CompAmmoUser>();
+        
+        public CompAmmoUser _compAmmo;
+        
+        public CompAmmoUser compAmmo
+        {
+            get
+            {
+                if(_compAmmo == null)
+                {
+                    _compAmmo = CompAmmo();
+                }
+                
+                return _compAmmo;
+            }
+        }
+
+
+        public CompFireModes CompFireModes() => this.parent.TryGetComp<CompFireModes>();
+        
+        public CompFireModes _compFireModes;
+        
+        public CompFireModes compFireModes
+        {
+            get
+            {
+                if(_compFireModes == null)
+                {
+                    _compFireModes = CompFireModes();
+                }
+                
+                return  _compFireModes;
+            }
+        }
+
+
+        public CompProperties_FireModes CompPropsFireModes() => this.parent.def.comps.Find(x => x is CompProperties_FireModes) as CompProperties_FireModes;
+        
+        public CompProperties_FireModes _compPropsFireModes;
+        
+        public CompProperties_FireModes compPropsFireModes
+        {
+            get
+            {
+                if(_compPropsFireModes == null)
+                {
+                    _compPropsFireModes = CompPropsFireModes();

```

Comment:
Normally, the public version is TitleCase, and the private version is camalCase.  Also, internal calculation is short enough to just inline it.

```
public CompProperties_FireModes CompPropsFireModes
        {
            get
            {
                if(_compPropsFireModes == null)
                {
                    _compPropsFireModes = this.parent.TryGetComp<CompFireModes>();
```